### PR TITLE
feat: improved types and exports

### LINKED ISSUE #960: Build failure with `exactOptionalPropertyTypes` enabled
**Is your feature request related to a problem? Please describe.**

Building the project using this library fails when `exactOptionalPropertyTypes` is enabled in project's `tsconfig.json` unless `skipLibCheck` is also set to `true`. `skipLibCheck` bypasses type checking for third-party libraries, compromising type safety.

**Describe the solution you'd like**
* Add the `exactOptionalPropertyTypes` property to `oclif/core`'s `tsconfig.json`.
* Make any necessary code changes to ensure successful build with this option enabled.

Example [rc/config/plugin.ts#L85](https://github.com/oclif/core/blob/e0c061a2cb2e0602b13bb6354fcf03f35ee9f6e7/src/config/plugin.ts#L85):

<table>
<thead>
<td>From</td>
<td>To</td>
</thead>
<tbody>
<tr>
<td>

```  
{ 
    parent: Plugin | undefined;    
}
```  

</td>
<td>

```
{
    parent?: Plugin;
}
```

</td>
</tr>
</tbody>
</table>

**Additional context**
* This change will improve type safety, leading to fewer potential runtime errors.
* It demonstrates adherence to best practices for TypeScript projects.

I would be happy to come up with a PR or collaborate on a potential solution.

### PR DESCRIPTION
**Pull request by mdonnalley**
- Improved `Interfaces.PJSON`
- Enable `exactOptionalPropertyTypes` (Fixes #960)
- Apply `target`/`identifier` logic to custom help classes
- theme can be a file path or json object
- Remove `default` property from configuration
- top-level exports:

```json
  "exports": {
    ".": "./lib/index.js",
    "./args": "./lib/args.js",
    "./command": "./lib/command.js",
    "./config": "./lib/config/index.js",
    "./errors": "./lib/errors/index.js",
    "./execute": "./lib/execute.js",
    "./flags": "./lib/flags.js",
    "./flush": "./lib/flush.js",
    "./handle": "./lib/errors/handle.js",
    "./help": "./lib/help/index.js",
    "./hooks": "./lib/interfaces/hooks.js",
    "./interfaces": "./lib/interfaces/index.js",
    "./logger": "./lib/logger.js",
    "./performance": "./lib/performance.js",
    "./run": "./lib/main.js",
    "./settings": "./lib/settings.js",
    "./util/ids": "./lib/util/ids.js",
    "./ux": "./lib/ux/index.js"
  },
```

@W-15470589@
@W-15470627@

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -82,7 +79,25 @@
     "argv"
   ],
   "license": "MIT",
-  "main": "./lib/index.js",
+  "exports": {
```
Comment (mshanemc) [inline package.json:82]: this includes all the ones I came here to check 😄

File: src/errors/errors/cli.ts [hunk 2]
```diff
@@ -22,12 +25,12 @@ export function addOclifExitCode(error: Record<string, any>, options?: {exit?: f
 }
 
 export class CLIError extends Error implements OclifError {
-  code?: string
+  code?: string | undefined
```
Comment (mshanemc) [inline src/errors/errors/cli.ts:28]: what's the value of having both `?` and the `| undefined` ?

File: src/config/config.ts [hunk 3]
```diff
@@ -83,8 +83,8 @@ export class Config implements IConfig {
   public arch!: ArchTypes
 
   public bin!: string
-  public binAliases?: string[]
-  public binPath?: string
+  public binAliases?: string[] | undefined
```
Comment (mshanemc) [inline src/config/config.ts:86]: should any of these be readonly?  conversely, do we want people mutating them all?

File: src/errors/errors/cli.ts (hunk 2 above)
Comment (mdonnalley) [reply src/errors/errors/cli.ts:28]: this https://github.com/oclif/core/issues/960

File: src/config/config.ts (hunk 3 above)
Comment (mdonnalley) [reply src/config/config.ts:86]: Can't be readonly because all these props are assigned in the `.load` method. Typescript doesn't like that

**Review by cristiand391 — APPROVED** [review]

**Pull request merged**